lib/Index/Index.cpp

```diff
@@ -468,26 +494,127 @@ class IndexSwiftASTWalker : public SourceEntityWalker {
   /// and \c b references to \c String and \c y.
   void beginBalancedDeclVisit(Decl *D) override {
     if (auto *PBD = dyn_cast<PatternBindingDecl>(D)) {
-      SmallVector<Decl *, 6> VDs;
+      PatternBindingPatterns Patterns;
+
       for (auto i : range(PBD->getNumPatternEntries())) {
+        PatternBindingPatternVarDecls VarDecls;
+
         PBD->getPattern(i)->forEachVariable(
-            [&](VarDecl *VD) { VDs.push_back(VD); });
+            [&](VarDecl *VD) { VarDecls.push_back(VD); });
+
+        if (VarDecls.size() > 0)
+          Patterns.push_back(std::move(VarDecls));
       }
 
-      if (VDs.size() > 0)
-        ContainerStack.push_back({std::move(VDs), D});
+      if (Patterns.size() > 0) {
+        ContainerStackEntry Entry{
+            PatternBinding, D, std::move(Patterns), {}, {}, {}, {}};
+        identifyPatternBindingSignposts(PBD, Entry);
+        ContainerStack.push_back(std::move(Entry));
+      }
     } else if (auto *AFD = dyn_cast<AbstractFunctionDecl>(D)) {
-      ContainerStack.push_back({{AFD}, D});
+      ContainerStack.push_back({Single, D, {}, {}, {}, {}, {}});
     }
   }
 
   void endBalancedDeclVisit(Decl *D) override {
     if (ContainerStack.empty())
       return;
-    if (ContainerStack.back().AncestorDecl == D)
+    if (ContainerStack.back().ContainerDecl == D)
       ContainerStack.pop_back();
   }
 
+  void identifyPatternBindingSignposts(PatternBindingDecl *PBD,
+                                       ContainerStackEntry &Entry) {
```

Comment:
Is there a reason we couldn't just associate a TypeRepr/Expr directly to VarDecl (if one) or Pattern (if on all) here?